### PR TITLE
clientv3: don't hold client lock while dialing

### LINKED ISSUE #5416: clientv3/integration: TestTxnWriteFail
```
=== RUN   TestTxnWriteFail
2016-05-20 06:05:40.712844 I | etcdserver: name = node2727028312981969055
2016-05-20 06:05:40.712954 I | etcdserver: data dir = /tmp/etcd551644292
2016-05-20 06:05:40.713065 I | etcdserver: member dir = /tmp/etcd551644292/member
2016-05-20 06:05:40.713129 I | etcdserver: heartbeat = 10ms
2016-05-20 06:05:40.713199 I | etcdserver: election = 100ms
2016-05-20 06:05:40.713284 I | etcdserver: snapshot count = 0
2016-05-20 06:05:40.713362 I | etcdserver: advertise client URLs = http://127.0.0.1:21532
2016-05-20 06:05:40.713470 I | etcdserver: initial advertise peer URLs = http://127.0.0.1:21531
2016-05-20 06:05:40.713543 I | etcdserver: name = node3807024210026896221
2016-05-20 06:05:40.713675 I | etcdserver: data dir = /tmp/etcd375866387
2016-05-20 06:05:40.713758 I | etcdserver: member dir = /tmp/etcd375866387/member
2016-05-20 06:05:40.713861 I | etcdserver: name = node8889113302011076440
2016-05-20 06:05:40.713989 I | etcdserver: data dir = /tmp/etcd550152278
2016-05-20 06:05:40.714083 I | etcdserver: member dir = /tmp/etcd550152278/member
2016-05-20 06:05:40.714165 I | etcdserver: heartbeat = 10ms
2016-05-20 06:05:40.714258 I | etcdserver: heartbeat = 10ms
2016-05-20 06:05:40.714356 I | etcdserver: election = 100ms
2016-05-20 06:05:40.714454 I | etcdserver: snapshot count = 0
2016-05-20 06:05:40.714549 I | etcdserver: initial cluster = node2727028312981969055=http://127.0.0.1:21531,node3807024210026896221=http://127.0.0.1:21533,node8889113302011076440=http://127.0.0.1:21535
2016-05-20 06:05:40.714694 I | etcdserver: election = 100ms
2016-05-20 06:05:40.714800 I | etcdserver: snapshot count = 0
2016-05-20 06:05:40.714926 I | etcdserver: advertise client URLs = http://127.0.0.1:21536
2016-05-20 06:05:40.715033 I | etcdserver: initial advertise peer URLs = http://127.0.0.1:21535
2016-05-20 06:05:40.715186 I | etcdserver: advertise client URLs = http://127.0.0.1:21534
2016-05-20 06:05:40.715342 I | etcdserver: initial advertise peer URLs = http://127.0.0.1:21533
2016-05-20 06:05:40.715490 I | etcdserver: initial cluster = node2727028312981969055=http://127.0.0.1:21531,node3807024210026896221=http://127.0.0.1:21533,node8889113302011076440=http://127.0.0.1:21535
2016-05-20 06:05:40.715625 I | etcdserver: initial cluster = node2727028312981969055=http://127.0.0.1:21531,node3807024210026896221=http://127.0.0.1:21533,node8889113302011076440=http://127.0.0.1:21535
2016-05-20 06:05:40.752477 I | etcdserver: starting member 6fac80429b1976f1 in cluster d909c0d7b220e389
2016-05-20 06:05:40.752782 I | raft: 6fac80429b1976f1 became follower at term 0
2016-05-20 06:05:40.752865 I | raft: newRaft 6fac80429b1976f1 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
2016-05-20 06:05:40.752915 I | raft: 6fac80429b1976f1 became follower at term 1
2016-05-20 06:05:40.752993 I | etcdserver: starting member e43be156984881f7 in cluster d909c0d7b220e389
2016-05-20 06:05:40.753215 I | raft: e43be156984881f7 became follower at term 0
2016-05-20 06:05:40.753306 I | raft: newRaft e43be156984881f7 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
2016-05-20 06:05:40.753363 I | raft: e43be156984881f7 became follower at term 1
2016-05-20 06:05:40.761315 I | etcdserver: starting member aea1f3ff542bdddf in cluster d909c0d7b220e389
2016-05-20 06:05:40.761671 I | raft: aea1f3ff542bdddf became follower at term 0
2016-05-20 06:05:40.761862 I | raft: newRaft aea1f3ff542bdddf [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
2016-05-20 06:05:40.761993 I | raft: aea1f3ff542bdddf became follower at term 1
2016-05-20 06:05:40.784310 I | etcdserver: set snapshot count to default 10000
2016-05-20 06:05:40.784490 I | etcdserver: starting server... [version: 3.0.0-beta.0+git, cluster version: to_be_decided]
2016-05-20 06:05:40.790548 I | etcdserver: set snapshot count to default 10000
2016-05-20 06:05:40.790820 I | etcdserver: starting server... [version: 3.0.0-beta.0+git, cluster version: to_be_decided]
2016-05-20 06:05:40.793215 N | etcdserver: added local member 6fac80429b1976f1 [http://127.0.0.1:21531] to cluster d909c0d7b220e389
2016-05-20 06:05:40.793997 N | etcdserver: added member aea1f3ff542bdddf [http://127.0.0.1:21535] to cluster d909c0d7b220e389
2016-05-20 06:05:40.795219 N | etcdserver: added member e43be156984881f7 [http://127.0.0.1:21533] to cluster d909c0d7b220e389
2016-05-20 06:05:40.800231 I | rafthttp: the connection with e43be156984881f7 became active
2016-05-20 06:05:40.801097 I | rafthttp: the connection with 6fac80429b1976f1 became active
2016-05-20 06:05:40.805315 N | etcdserver: added member 6fac80429b1976f1 [http://127.0.0.1:21531] to cluster d909c0d7b220e389
2016-05-20 06:05:40.805648 I | rafthttp: the connection with aea1f3ff542bdddf became active
2016-05-20 06:05:40.805748 I | etcdserver: set snapshot count to default 10000
2016-05-20 06:05:40.806592 I | etcdserver: starting server... [version: 3.0.0-beta.0+git, cluster version: to_be_decided]
2016-05-20 06:05:40.806818 I | rafthttp: the connection with 6fac80429b1976f1 became active
2016-05-20 06:05:40.808301 N | etcdserver: added member aea1f3ff542bdddf [http://127.0.0.1:21535] to cluster d909c0d7b220e389
2016-05-20 06:05:40.809172 N | etcdserver: added local member e43be156984881f7 [http://127.0.0.1:21533] to cluster d909c0d7b220e389
2016-05-20 06:05:40.809874 N | etcdserver: added member 6fac80429b1976f1 [http://127.0.0.1:21531] to cluster d909c0d7b220e389
2016-05-20 06:05:40.810303 N | etcdserver: added local member aea1f3ff542bdddf [http://127.0.0.1:21535] to cluster d909c0d7b220e389
2016-05-20 06:05:40.810741 N | etcdserver: added member e43be156984881f7 [http://127.0.0.1:21533] to cluster d909c0d7b220e389
2016-05-20 06:05:40.811601 I | rafthttp: the connection with aea1f3ff542bdddf became active
2016-05-20 06:05:40.813333 I | rafthttp: the connection with e43be156984881f7 became active
2016-05-20 06:05:40.843427 I | raft: 6fac80429b1976f1 is starting a new election at term 1
2016-05-20 06:05:40.843549 I | raft: 6fac80429b1976f1 became candidate at term 2
2016-05-20 06:05:40.843612 I | raft: 6fac80429b1976f1 received vote from 6fac80429b1976f1 at term 2
2016-05-20 06:05:40.843685 I | raft: 6fac80429b1976f1 [logterm: 1, index: 3] sent vote request to aea1f3ff542bdddf at term 2
2016-05-20 06:05:40.843754 I | raft: 6fac80429b1976f1 [logterm: 1, index: 3] sent vote request to e43be156984881f7 at term 2
2016-05-20 06:05:40.844951 I | raft: aea1f3ff542bdddf [term: 1] received a MsgVote message with higher term from 6fac80429b1976f1 [term: 2]
2016-05-20 06:05:40.845043 I | raft: e43be156984881f7 [term: 1] received a MsgVote message with higher term from 6fac80429b1976f1 [term: 2]
2016-05-20 06:05:40.845192 I | raft: e43be156984881f7 became follower at term 2
2016-05-20 06:05:40.845268 I | raft: aea1f3ff542bdddf became follower at term 2
2016-05-20 06:05:40.845330 I | raft: e43be156984881f7 [logterm: 1, index: 3, vote: 0] voted for 6fac80429b1976f1 [logterm: 1, index: 3] at term 2
2016-05-20 06:05:40.845416 I | raft: aea1f3ff542bdddf [logterm: 1, index: 3, vote: 0] voted for 6fac80429b1976f1 [logterm: 1, index: 3] at term 2
2016-05-20 06:05:40.845957 I | raft: 6fac80429b1976f1 received vote from aea1f3ff542bdddf at term 2
2016-05-20 06:05:40.846021 I | raft: 6fac80429b1976f1 [quorum:2] has received 2 votes and 0 vote rejections
2016-05-20 06:05:40.846095 I | raft: 6fac80429b1976f1 became leader at term 2
2016-05-20 06:05:40.846158 I | raft: raft.node: 6fac80429b1976f1 elected leader 6fac80429b1976f1 at term 2
2016-05-20 06:05:40.846695 I | raft: raft.node: aea1f3ff542bdddf elected leader 6fac80429b1976f1 at term 2
2016-05-20 06:05:40.847134 I | raft: raft.node: e43be156984881f7 elected leader 6fac80429b1976f1 at term 2
2016-05-20 06:05:40.849514 I | etcdserver: published {Name:node2727028312981969055 ClientURLs:[http://127.0.0.1:21532]} to cluster d909c0d7b220e389
2016-05-20 06:05:40.850589 I | etcdserver: published {Name:node8889113302011076440 ClientURLs:[http://127.0.0.1:21536]} to cluster d909c0d7b220e389
2016-05-20 06:05:40.850991 I | etcdserver: published {Name:node3807024210026896221 ClientURLs:[http://127.0.0.1:21534]} to cluster d909c0d7b220e389
2016-05-20 06:05:40.851059 I | etcdserver: setting up the initial cluster version to 3.0
2016-05-20 06:05:40.851975 N | membership: set the initial cluster version to 3.0
2016-05-20 06:05:40.852442 N | membership: set the initial cluster version to 3.0
2016-05-20 06:05:40.852597 N | membership: set the initial cluster version to 3.0
2016-05-20 06:05:40.868305 E | rafthttp: failed to read aea1f3ff542bdddf on stream Message (net/http: request canceled)
2016-05-20 06:05:40.868385 I | rafthttp: the connection with aea1f3ff542bdddf became inactive
2016-05-20 06:05:40.953143 I | raft: aea1f3ff542bdddf is starting a new election at term 2
2016-05-20 06:05:40.953352 I | raft: aea1f3ff542bdddf became candidate at term 3
2016-05-20 06:05:40.953458 I | raft: aea1f3ff542bdddf received vote from aea1f3ff542bdddf at term 3
2016-05-20 06:05:40.953606 I | raft: aea1f3ff542bdddf [logterm: 2, index: 8] sent vote request to 6fac80429b1976f1 at term 3
2016-05-20 06:05:40.953747 I | raft: aea1f3ff542bdddf [logterm: 2, index: 8] sent vote request to e43be156984881f7 at term 3
2016-05-20 06:05:40.953851 I | raft: raft.node: aea1f3ff542bdddf lost leader 6fac80429b1976f1 at term 3
2016-05-20 06:05:40.954786 I | raft: e43be156984881f7 [term: 2] received a MsgVote message with higher term from aea1f3ff542bdddf [term: 3]
2016-05-20 06:05:40.954943 I | raft: e43be156984881f7 became follower at term 3
2016-05-20 06:05:40.955097 I | raft: e43be156984881f7 [logterm: 2, index: 8, vote: 0] voted for aea1f3ff542bdddf [logterm: 2, index: 8] at term 3
2016-05-20 06:05:40.955206 I | raft: raft.node: e43be156984881f7 lost leader 6fac80429b1976f1 at term 3
2016-05-20 06:05:40.955894 I | raft: aea1f3ff542bdddf received vote from e43be156984881f7 at term 3
2016-05-20 06:05:40.955994 I | raft: aea1f3ff542bdddf [quorum:2] has received 2 votes and 0 vote rejections
2016-05-20 06:05:40.956122 I | raft: aea1f3ff542bdddf became leader at term 3
2016-05-20 06:05:40.956243 I | raft: raft.node: aea1f3ff542bdddf elected leader aea1f3ff542bdddf at term 3
2016-05-20 06:05:40.957164 I | raft: raft.node: e43be156984881f7 elected leader aea1f3ff542bdddf at term 3
2016-05-20 06:05:40.958724 E | rafthttp: failed to dial 6fac80429b1976f1 on stream MsgApp v2 (dial tcp 127.0.0.1:21531: getsockopt: connection refused)
2016-05-20 06:05:40.958848 I | rafthttp: the connection with 6fac80429b1976f1 became inactive
2016-05-20 06:05:40.958978 W | etcdserver: failed to reach the peerURL(http://127.0.0.1:21531) of member 6fac80429b1976f1 (Get http://127.0.0.1:21531/version: dial tcp 127.0.0.1:21531: getsockopt: connection refused)
2016-05-20 06:05:40.959147 W | etcdserver: cannot get the version of member 6fac80429b1976f1 (Get http://127.0.0.1:21531/version: dial tcp 127.0.0.1:21531: getsockopt: connection refused)
2016-05-20 06:05:40.972053 E | rafthttp: failed to dial 6fac80429b1976f1 on stream MsgApp v2 (dial tcp 127.0.0.1:21531: getsockopt: connection refused)
2016-05-20 06:05:40.972214 I | rafthttp: the connection with 6fac80429b1976f1 became inactive
2016-05-20 06:05:41.293534 I | v2http: enabled capabilities for version 3.0.0
2016-05-20 06:05:41.311396 I | v2http: enabled capabilities for version 3.0.0
2016-05-20 06:05:44.963000 W | etcdserver: failed to reach the peerURL(http://127.0.0.1:21531) of member 6fac80429b1976f1 (Get http://127.0.0.1:21531/version: dial tcp 127.0.0.1:21531: getsockopt: connection refused)
2016-05-20 06:05:44.963277 W | etcdserver: cannot get the version of member 6fac80429b1976f1 (Get http://127.0.0.1:21531/version: dial tcp 127.0.0.1:21531: getsockopt: connection refused)
2016-05-20 06:05:48.967789 W | etcdserver: failed to reach the peerURL(http://127.0.0.1:21531) of member 6fac80429b1976f1 (Get http://127.0.0.1:21531/version: dial tcp 127.0.0.1:21531: getsockopt: connection refused)
2016-05-20 06:05:48.967959 W | etcdserver: cannot get the version of member 6fac80429b1976f1 (Get http://127.0.0.1:21531/version: dial tcp 127.0.0.1:21531: getsockopt: connection refused)
2016-05-20 06:05:52.970492 W | etcdserver: failed to reach the peerURL(http://127.0.0.1:21531) of member 6fac80429b1976f1 (Get http://127.0.0.1:21531/version: dial tcp 127.0.0.1:21531: getsockopt: connection refused)
2016-05-20 06:05:52.970736 W | etcdserver: cannot get the version of member 6fac80429b1976f1 (Get http://127.0.0.1:21531/version: dial tcp 127.0.0.1:21531: getsockopt: connection refused)
--- FAIL: TestTxnWriteFail (12.42s)
	txn_test.go:78: timed out waiting for txn to fail
```

### PR DESCRIPTION
**Pull request by heyitsanthony**
Causes async reconnect to block while the client is dialing.

Fixes #5416

### CODE REVIEW

**Comment by gyuho** [comment]
lgtm. defer to @xiang90 

THanks.

**Comment by heyitsanthony** [comment]
Without patch on remoteClient.reconnect:
```
reconnect get lock 2016-06-02 11:34:40.675922611 -0700 PDT
reconnect got lock 2016-06-02 11:34:40.675966745 -0700 PDT
reconnect done 2016-06-02 11:34:45.550275821 -0700 PDT

reconnect get lock 2016-06-02 11:34:52.456397151 -0700 PDT
reconnect got lock 2016-06-02 11:34:52.456473547 -0700 PDT
reconnect done 2016-06-02 11:34:54.89495976 -0700 PDT

reconnect get lock 2016-06-02 11:35:02.037813784 -0700 PDT
reconnect got lock 2016-06-02 11:35:02.03786161 -0700 PDT
reconnect done 2016-06-02 11:35:04.910668863 -0700 PDT
```

With patch:
```
reconnect get lock 2016-06-02 12:06:56.688914062 -0700 PDT
reconnect got lock 2016-06-02 12:06:56.688961965 -0700 PDT
reconnect done 2016-06-02 12:06:56.689264578 -0700 PDT

reconnect get lock 2016-06-02 12:07:01.705181582 -0700 PDT
reconnect got lock 2016-06-02 12:07:01.715701388 -0700 PDT
reconnect done 2016-06-02 12:07:01.715753236 -0700 PDT

reconnect get lock 2016-06-02 12:07:06.859568984 -0700 PDT
reconnect got lock 2016-06-02 12:07:06.859619072 -0700 PDT
reconnect done 2016-06-02 12:07:06.859781505 -0700 PDT
```

File: clientv3/client.go [hunk 1]
```diff
@@ -298,14 +297,20 @@ func (c *Client) retryConnection(err error) (newConn *grpc.ClientConn, dialErr e
 	}
 	if c.cancel == nil {
 		// client has called Close() so don't try to dial out
+		c.mu.Unlock()
 		return nil, c.ctx.Err()
 	}
+	c.mu.Unlock()
 
-	c.conn, dialErr = c.cfg.retryDialer(c)
+	newConn, dialErr = c.cfg.retryDialer(c)
```
Comment (xiang90) [inline clientv3/client.go]: here we might leave c.conn to nil while reconnecting, will this be an issue?

Comment (heyitsanthony) [inline clientv3/client.go]: fixed

**Comment by xiang90** [comment]
lgtm

**Comment by heyitsanthony** [comment]
reworked the client so it'll always have a non-nil (but sometimes closed) ActiveConnection and to use lastConnErr to signify the last dial failed so there doesn't have to be so much nil handling. ptal

File: clientv3/client.go [hunk 2]
```diff
@@ -282,30 +296,39 @@ func (c *Client) ActiveConnection() *grpc.ClientConn {
 }
 
 // retryConnection establishes a new connection
-func (c *Client) retryConnection(err error) (newConn *grpc.ClientConn, dialErr error) {
+func (c *Client) retryConnection(err error) {
+	oldconn := c.conn
+	defer func() {
```
Comment (xiang90) [inline clientv3/client.go]: can we put the defer after c.mu.lock? it seems strange to have c.mu.unlock before c.mu.lock

Comment (xiang90) [inline clientv3/client.go]: hm... it is strange anyway. this unlock is not pair with the first lock. probably some comments?

**Comment by heyitsanthony** [comment]
fixed up /cc @xiang90

**Comment by xiang90** [comment]
lgtm

**Pull request merged**